File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteContactRepository.java
package de.qabel.core.repository.sqlite;

import de.qabel.core.StringUtils;
import de.qabel.core.config.Contact;
import de.qabel.core.config.Contacts;
import de.qabel.core.config.Identity;
import de.qabel.core.config.factory.DefaultContactFactory;
import de.qabel.core.repository.ContactRepository;
import de.qabel.core.repository.EntityManager;
import de.qabel.core.repository.exception.EntityNotFoundException;
import de.qabel.core.repository.exception.PersistenceException;
import de.qabel.core.repository.sqlite.hydrator.ContactHydrator;
import de.qabel.core.repository.sqlite.hydrator.DropURLHydrator;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class SqliteContactRepository extends AbstractSqliteRepository<Contact> implements ContactRepository {
    public static final String TABLE_NAME = "contact";
    private final SqliteDropUrlRepository dropUrlRepository;

    public SqliteContactRepository(ClientDatabase database, EntityManager em) {
        this(
            database,
            new ContactHydrator(
                em,
                new DefaultContactFactory(),
                new SqliteDropUrlRepository(database, new DropURLHydrator())
            ),
            new SqliteDropUrlRepository(database, new DropURLHydrator())
        );
    }

    public SqliteContactRepository(ClientDatabase database, Hydrator<Contact> hydrator, SqliteDropUrlRepository dropUrlRepository) {
        super(database, hydrator, TABLE_NAME);
        this.dropUrlRepository = dropUrlRepository;
    }

    public Contact find(Integer id) throws PersistenceException, EntityNotFoundException {
        return findBy("id=?", id);
    }

    @Override
    public synchronized Contacts find(Identity identity) throws PersistenceException {
        Contacts contacts = new Contacts(identity);

        try (PreparedStatement statement = database.prepare(
            "SELECT " + StringUtils.join(",", hydrator.getFields("c")) + " " +
            "FROM contact c " +
            "JOIN identity_contacts ic ON (c.id = ic.contact_id) " +
            "JOIN identity i ON (ic.identity_id = i.id) " +
            "JOIN contact c2 ON (i.contact_id = c2.id) " +
            "WHERE c2.publicKey = ?"
        )) {
            statement.setString(1, identity.getKeyIdentifier());
            try (ResultSet resultSet = statement.executeQuery()) {
                for (Contact c : hydrator.hydrateAll(resultSet)) {
                    contacts.put(c);
                }
            }
        } catch (SQLException e) {
            throw new PersistenceException("failed to load contacts for " + identity, e);
        }

        return contacts;
    }

    @Override
    public synchronized void save(Contact contact, Identity identity) throws PersistenceException {
        try {
            if (contact.getId() == 0 || !exists(contact)) {
                insert(contact, identity);
            } else {
                update(contact, identity);
            }
            dropUrlRepository.delete(contact);
            dropUrlRepository.store(contact);
            hydrator.recognize(contact);
        } catch (SQLException e) {
            throw new PersistenceException("failed to save contact: " + e.getMessage(), e);
        }
    }

    private boolean exists(Contact contact) throws SQLException {
        try (PreparedStatement statement = database.prepare("SELECT id FROM contact WHERE id = ?")) {
            statement.setInt(1, contact.getId());
            try (ResultSet resultSet = statement.executeQuery()) {
                return resultSet.next();
            }
        }
    }

    private void insert(Contact contact, Identity identity) throws SQLException {
        try {
            Contact existing = findBy("publicKey=?", contact.getKeyIdentifier());
            contact.setId(existing.getId());
        } catch (PersistenceException | EntityNotFoundException e) {
            try (PreparedStatement statement = database.prepare(
                "INSERT INTO contact (publicKey, alias, phone, email) VALUES (?, ?, ?, ?)"
            )) {
                int i = 1;
                statement.setString(i++, contact.getKeyIdentifier());
                statement.setString(i++, contact.getAlias());
                statement.setString(i++, contact.getPhone());
                statement.setString(i++, contact.getEmail());
                statement.execute();
                try (ResultSet keys = statement.getGeneratedKeys()) {
                    keys.next();
                    contact.setId(keys.getInt(1));
                }
            }
        }
        insertConnection(contact, identity);
    }

    private void insertConnection(Contact contact, Identity identity) throws SQLException {
        try (PreparedStatement statement = database.prepare(
            "INSERT OR IGNORE INTO identity_contacts (identity_id, contact_id) VALUES (?, ?)"
        )) {
            int i = 1;
            statement.setInt(i++, identity.getId());
            statement.setInt(i++, contact.getId());
            statement.execute();
        }
    }

    private void update(Contact contact, Identity identity) throws SQLException {
        try (PreparedStatement statement = database.prepare(
            "UPDATE contact SET publicKey=?, alias=?, phone=?, email=? WHERE id=?"
        )) {
            int i = 1;
            statement.setString(i++, contact.getKeyIdentifier());
            statement.setString(i++, contact.getAlias());
            statement.setString(i++, contact.getPhone());
            statement.setString(i++, contact.getEmail());
            statement.setInt(i++, contact.getId());
            statement.execute();
        }
        insertConnection(contact, identity);
    }

    @Override
    public synchronized void delete(Contact contact, Identity identity) throws PersistenceException, EntityNotFoundException {
        try {
            try (PreparedStatement statement = database.prepare(
                "DELETE FROM identity_contacts WHERE contact_id = ? AND identity_id = ?"
            )) {
                int i = 1;
                statement.setInt(i++, contact.getId());
                statement.setInt(i++, identity.getId());
                statement.execute();
                if (statement.getUpdateCount() != 1) {
                    throw new EntityNotFoundException(
                            "Contact "+contact.getAlias() +" for identity "
                                    + identity.getAlias() +" not found");
                }
            }
            try (PreparedStatement statement = database.prepare(
                "DELETE FROM contact WHERE id = ? AND NOT EXISTS (" +
                    "SELECT contact_id FROM identity_contacts WHERE contact_id = ? LIMIT 1" +
                ")"
            )) {
                statement.setInt(1, contact.getId());
                statement.setInt(2, contact.getId());
                statement.execute();
            }
        } catch (SQLException e) {
            throw new PersistenceException("failed to delete contact", e);
        }
    }

    @Override
    public synchronized Contact findByKeyId(Identity identity, String keyId) throws EntityNotFoundException {
        try {
            try (PreparedStatement statement = database.prepare(
                "SELECT " + StringUtils.join(",", hydrator.getFields("c")) + " FROM " + TABLE_NAME + " c " +
                    "JOIN identity_contacts ic ON (c.id = ic.contact_id) " +
                    "JOIN identity i ON (i.id = ic.identity_id) " +
                    "JOIN contact c2 ON (c2.id = i.contact_id) " +
                    "WHERE c2.publicKey = ? AND c.publicKey = ? " +
                    "LIMIT 1"
            )) {
                statement.setString(1, identity.getKeyIdentifier());
                statement.setString(2, keyId);
                try (ResultSet results = statement.executeQuery()) {
                    if (!results.next()) {
                        throw new EntityNotFoundException(
                            "no contact found for identity '" + identity.getAlias() + "' and key '" + keyId + "'");
                    }
                    return hydrator.hydrateOne(results);
                }
            }
        } catch (SQLException e) {
            throw new EntityNotFoundException("exception while searching contact: " + e.getMessage(), e);
        }
    }
}

File: core/src/main/java/de/qabel/core/repository/sqlite/SqliteContactRepository.kt
package de.qabel.core.repository.sqlite

import de.qabel.core.config.*
import de.qabel.core.contacts.ContactData
import de.qabel.core.extensions.findById
import de.qabel.core.logging.QabelLog
import de.qabel.core.repository.ContactRepository
import de.qabel.core.repository.DropUrlRepository
import de.qabel.core.repository.EntityManager
import de.qabel.core.repository.IdentityRepository
import de.qabel.core.repository.exception.EntityExistsException
import de.qabel.core.repository.exception.EntityNotFoundException
import de.qabel.core.repository.framework.BaseRepository
import de.qabel.core.repository.framework.QueryBuilder
import de.qabel.core.repository.framework.ResultAdapter
import de.qabel.core.repository.sqlite.hydrator.ContactAdapter
import de.qabel.core.repository.sqlite.hydrator.DropURLHydrator
import de.qabel.core.repository.sqlite.hydrator.IntResultAdapter
import de.qabel.core.repository.sqlite.schemas.ContactDB
import de.qabel.core.repository.sqlite.schemas.ContactDB.ContactDropUrls
import de.qabel.core.repository.sqlite.schemas.ContactDB.IdentityContacts
import de.qabel.core.util.DefaultHashMap
import java.sql.ResultSet
import java.util.*


class SqliteContactRepository(db: ClientDatabase, em: EntityManager,
                              dropUrlRepository: DropUrlRepository = SqliteDropUrlRepository(db, DropURLHydrator()),
                              private val identityRepository: IdentityRepository = SqliteIdentityRepository(db, em)) :
    BaseRepository<Contact>(ContactDB, ContactAdapter(dropUrlRepository), db, em), ContactRepository, QabelLog, EntityObservable by SimpleEntityObservable() {

    override fun find(id: Int): Contact = findById(id)

    override fun find(identity: Identity): Contacts =
        with(createEntityQuery()) {
            joinIdentityContacts(this)
            whereAndEquals(IdentityContacts.IDENTITY_ID, identity.id)
            val resultList = getResultList<Contact>(this)
            return Contacts(identity).apply {
                resultList.forEach { put(it) }
            }
        }

    override fun save(contact: Contact, identity: Identity) {
        val exists = exists(contact)
        val affectedContact = if (contact.id == 0 && exists) {
            val existingContact = findByKeyId(contact.keyIdentifier)
            existingContact.alias = contact.alias
            existingContact.email = contact.email
            existingContact.phone = contact.phone
            existingContact
        } else {
            contact
        }

        if (affectedContact.id == 0 || !exists) {
            persist(affectedContact)
        } else {
            update(affectedContact)
        }

        addIdentityConnection(affectedContact, identity)
        notifyObservers()
    }

    override fun update(contact: Contact, activeIdentities: List<Identity>) {
        update(contact)
        removeIdentityConnections(contact)
        if (activeIdentities.isNotEmpty()) {
            addIdentityConnections(contact, activeIdentities)
        }
        notifyObservers()
    }

    override fun persist(contact: Contact, identities: List<Identity>) {
        if (exists(contact)) {
            throw EntityExistsException("Contact already exists!")
        }
        persist(contact)
        if (identities.isNotEmpty()) {
            addIdentityConnections(contact, identities)
        }
        notifyObservers()
    }

    override fun persist(model: Contact) {
        super.persist(model)
        model.dropUrls.forEach { saveManyToMany(ContactDropUrls.CONTACT_ID, model.id, ContactDropUrls.DROP_URL, it.toString()) }
        info("Contact ${model.alias} persisted with id ${model.id}")
    }

    override fun update(model: Contact) {
        super.update(model)
        dropAllManyToMany(ContactDropUrls.CONTACT_ID, model.id)
        model.dropUrls.forEach { saveManyToMany(ContactDropUrls.CONTACT_ID, model.id, ContactDropUrls.DROP_URL, it.toString()) }
        info("Contact ${model.alias} (${model.id}) updated ")
    }

    override fun delete(contact: Contact, identity: Identity) {
        val contactIdentities = getIdentityConnections(contact)
        removeIdentityConnection(contact, identity)
        if (contactIdentities.size == 1 && contactIdentities.first() == identity.id) {
            delete(contact.id)
        }
    }

    override fun delete(contact: Contact) = delete(contact.id)

    override fun delete(id: Int) {
        val contact = findById(id)
        removeIdentityConnections(contact)
        dropAllManyToMany(ContactDropUrls.CONTACT_ID, id)
        super.delete(id)
        info("Contact ${contact.alias} ($id) deleted")
        notifyObservers()
    }

    private fun joinIdentityContacts(queryBuilder: QueryBuilder) =
        queryBuilder.innerJoin(IdentityContacts.TABLE, IdentityContacts.TABLE_ALIAS,
            IdentityContacts.CONTACT_ID.exp(), ContactDB.ID.exp())

    private fun addIdentityConnection(contact: Contact, identity: Identity) =
        saveManyToMany(IdentityContacts.IDENTITY_ID, identity.id, IdentityContacts.CONTACT_ID, contact.id)

    private fun addIdentityConnections(contact: Contact, identities: List<Identity>) =
        saveManyToMany(IdentityContacts.CONTACT_ID, contact.id, IdentityContacts.IDENTITY_ID,
            *identities.map { it.id }.toTypedArray())

    private fun removeIdentityConnection(contact: Contact, identity: Identity) =
        dropManyToMany(IdentityContacts.IDENTITY_ID, identity.id, IdentityContacts.CONTACT_ID, contact.id)

    private fun removeIdentityConnections(contact: Contact) =
        dropAllManyToMany(IdentityContacts.CONTACT_ID, contact.id)

    private fun getIdentityConnections(contact: Contact) =
        findManyToMany(IdentityContacts.CONTACT_ID, contact.id, IdentityContacts.IDENTITY_ID, IntResultAdapter())


    override fun findByKeyId(identity: Identity, keyId: String): Contact =
        with(createEntityQuery()) {
            joinIdentityContacts(this)
            whereAndEquals(IdentityContacts.IDENTITY_ID, identity.id)
            whereAndEquals(ContactDB.PUBLIC_KEY, keyId)
            return this@SqliteContactRepository.getSingleResult(this)
        }

    override fun findByKeyId(keyId: String): Contact =
        with(createEntityQuery()) {
            whereAndEquals(ContactDB.PUBLIC_KEY, keyId)
            return this@SqliteContactRepository.getSingleResult(this)
        }

    override fun exists(contact: Contact): Boolean = try {
        findByKeyId(contact.keyIdentifier); true
    } catch(ex: EntityNotFoundException) {
        false
    }

    override fun findContactWithIdentities(keyId: String): ContactData {
        val contact = findByKeyId(keyId)
        val identities = identityRepository.findAll()
        val identityKeys = getIdentityConnections(contact)
        return ContactData(contact, identities.entities.filter { identityKeys.contains(it.id) },
            identities.contains(contact.keyIdentifier))
    }

    private fun findIdentityIds(contacts: List<Contact>): Map<Int, List<Int>>
        = DefaultHashMap<Int, MutableList<Int>>({ LinkedList<Int>() }).apply {

        findManyToMany(IdentityContacts.CONTACT_ID, IdentityContacts.IDENTITY_ID,
            object : ResultAdapter<Unit> {
                override fun hydrateOne(resultSet: ResultSet, entityManager: EntityManager, detached : Boolean) {
                    val contactId = resultSet.getInt(1)
                    val identityId = resultSet.getInt(2)
                    getOrDefault(contactId).add(identityId)
                }
            }, contacts.map { it.id })
    }

    override fun findWithIdentities(searchString: String, status: List<Contact.ContactStatus>, excludeIgnored: Boolean): Collection<ContactData> {
        info("findWithIdentities with filters")
        val contacts = with(createEntityQuery()) {
            //Exclude identities
            leftJoin(ContactDB.IdentityJoin.TABLE, ContactDB.IdentityJoin.TABLE_ALIAS,
                ContactDB.IdentityJoin.CONTACT_ID.exp(), relation.ID.exp())
            whereAndNull(ContactDB.IdentityJoin.CONTACT_ID)

            if (excludeIgnored) {
                whereAndEquals(ContactDB.IGNORED, false)
            }
            whereAndIn(ContactDB.STATUS, status.map { it.status })
            if (!searchString.isEmpty()) {
                whereAndLowerEquals(searchString, ContactDB.ALIAS, ContactDB.NICKNAME,
                    ContactDB.EMAIL, ContactDB.PHONE)
            }
            orderBy(ContactDB.ALIAS.exp())
            getResultList<Contact>(this)
        }
        val identities = identityRepository.findAll()
        val contactIdentities = findIdentityIds(contacts)
        return contacts.map {
            ContactData(it, contactIdentities[it.id]!!.map {
                identities.entities.findById(it)!!
            }, identities.contains(it.keyIdentifier))
        }
    }
}
